src/hotspot/share/opto/node.cpp

```diff
@@ -1601,6 +1602,199 @@ Node* old_root() {
+  Node_List _worklist;
+};
+
+// BFS traverse all reachable nodes from start, call callback on them
+template <typename Callback>
+void visit_nodes(Node* start, Callback callback, bool traverse_output, bool only_ctrl) {
+  UniqueMixedNodeList worklist;
+  worklist.add(start);
+  for (uint i = 0; i < worklist.size(); i++) {
+    Node* n = worklist[i];
+    callback(n);
+    for (uint i = 0; i < n->len(); i++) {
+      if (!only_ctrl || n->is_Region() || (n->Opcode() == Op_Root) || (i == TypeFunc::Control)) {
+        // If only_ctrl is set: Add regions, the root node, or control inputs only
+        worklist.add(n->in(i));
+      }
+    }
+    if (traverse_output && !only_ctrl) {
+      for (uint i = 0; i < n->outcnt(); i++) {
+        worklist.add(n->raw_out(i));
+      }
+    }
+  }
+}
+
+// BFS traverse from start, return node with idx
+Node* find_node_by_idx(Node* start, uint idx, bool traverse_output, bool only_ctrl) {
+  Node* result = nullptr;
+  auto callback = [&] (Node* n) {
+    if (n->_idx == idx) {
+      if (result != nullptr) {
+        tty->print("find_node_by_idx: " INTPTR_FORMAT " and " INTPTR_FORMAT " both have idx==%d\n",
+          (uintptr_t)result, (uintptr_t)n, idx);
+      }
+      result = n;
+    }
+  };
+  visit_nodes(start, callback, traverse_output, only_ctrl);
+  return result;
+}
+
+// find needle in haystack, case insensitive
+// custom implementation of strcasestr, as it is not available on windows
+const char* strstr_nocase(const char* haystack, const char* needle) {
+  if (needle[0] == '\0') {
+    return haystack; // empty needle matches with anything
+  }
+  for (size_t i = 0; haystack[i] != '\0'; i++) {
+    bool matches = true;
+    for (size_t j = 0; needle[j] != '\0'; j++) {
+      if (haystack[i + j] == '\0') {
+        return nullptr; // hit end of haystack, abort
+      }
+      if (tolower(haystack[i + j]) != tolower(needle[j])) {
+        matches = false;
+        break; // abort, try next i
+      }
+    }
+    if (matches) {
+      return &haystack[i]; // all j were ok for this i
+    }
+  }
+  return nullptr; // no i was a match
+}
+
+// check if str matches the star_pattern
+// eg. str "_abc____def__" would match pattern "abc*def"
+// the matching is case insensitive
+bool is_star_match(const char* star_pattern, const char* str) {
+  const int N = 1000;
+  char pattern[N]; // copy pattern into this to ensure null termination
+  jio_snprintf(pattern, N, "%s", star_pattern);// ensures null termination
+  char buf[N]; // copy parts of pattern into this
+  const char* s = str;
+  const char* r = &pattern[0]; // cast array to char*
+  while (strlen(r) > 0) {
+    // find next section in pattern
+    const char* r_end = strstr(r, "*");
+    const char* r_part = r;
+    if (r_end != nullptr) { // copy part into buffer
+      size_t r_part_len = r_end-r;
+      strncpy(buf, r, r_part_len);
+      buf[r_part_len] = '\0'; // end of string
+      r_part = &buf[0]; // cast array to char*
```

Comment:
```suggestion
      r_part = buf;
```

Reply:
checking if I can do that, remember I had some issues with a platform.

Reply:
Tests passed fine, so I will do as you suggested. Thanks for the feedback.